File: backup_php/public/api/panel/servicos.php
<?php
ini_set('display_errors', 0);
error_reporting(0);

// Carrega o autoloader do Composer
require_once __DIR__ . '/../../../vendor/autoload.php';

// Carrega as variáveis de ambiente
$envPath = __DIR__ . '/../../../.env';
if (file_exists($envPath)) {
    $dotenv = Dotenv\Dotenv::createImmutable(__DIR__ . '/../../../');
    $dotenv->load();
}

// Usa a classe Database para conexão
use App\Config\Database;

try {
    $pdo = Database::getConnection();
} catch (PDOException $e) {
    http_response_code(500);
    echo json_encode(['error' => 'Erro na conexão com banco de dados']);
    exit;
}


header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type');

if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') {
    exit(0);
}

try {
    
    if ($_SERVER['REQUEST_METHOD'] === 'GET') {
        // Se há um ID na query string, buscar serviço específico
        if (isset($_GET['id'])) {
            error_log("Buscando serviço com ID: " . $_GET['id']);
            
            $stmt = $pdo->prepare("SELECT * FROM servicos_clinica WHERE id = ?");
            $stmt->execute([$_GET['id']]);
            $servico = $stmt->fetch(PDO::FETCH_ASSOC);
            
            error_log("Serviço encontrado: " . ($servico ? 'sim' : 'não'));
            if ($servico) {
                error_log("Dados do serviço: " . json_encode($servico));
            }
            
            if ($servico) {
                echo json_encode([
                    'success' => true,
                    'servico' => $servico
                ]);
            } else {
                echo json_encode([
                    'success' => false,
                    'message' => 'Serviço não encontrado'
                ]);
            }
        } else {
            // Buscar todos os serviços
            $stmt = $pdo->prepare("SELECT * FROM servicos_clinica ORDER BY nome");
            $stmt->execute();
            $servicos = $stmt->fetchAll(PDO::FETCH_ASSOC);
            
            echo json_encode([
                'success' => true,
                'servicos' => $servicos
            ]);
        }
        
    } elseif ($_SERVER['REQUEST_METHOD'] === 'POST') {
        // Criar novo serviço
        $input = json_decode(file_get_contents('php://input'), true);
        
        if (!$input || !isset($input['nome'])) {
            throw new Exception('Nome do serviço é obrigatório');
        }
        
        $stmt = $pdo->prepare("
            INSERT INTO servicos_clinica (
                nome, descricao, valor, ativo, 
                palavras_chave, categoria, observacoes, 
                preparo_necessario, anestesia_tipo, local_realizacao
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        ");
        
        $stmt->execute([
            $input['nome'],
            $input['descricao'] ?? null,
            $input['valor'] === '' ? null : ($input['valor'] ?? null),
            $input['ativo'] ?? 1,
            $input['palavras_chave'] ?? null,
            $input['categoria'] ?? null,
            $input['observacoes'] ?? null,
            $input['preparo_necessario'] ?? null,
            $input['anestesia_tipo'] ?? null,
            $input['local_realizacao'] ?? null
        ]);
        
        echo json_encode([
            'success' => true,
            'message' => 'Serviço criado com sucesso',
            'id' => $pdo->lastInsertId()
        ]);
        
    } elseif ($_SERVER['REQUEST_METHOD'] === 'PUT') {
        // Atualizar serviço (suporte a atualizações parciais)
        $input = json_decode(file_get_contents('php://input'), true);
        
        if (!$input || !isset($input['id'])) {
            throw new Exception('ID é obrigatório');
        }
        
        // Construir query dinamicamente baseada nos campos fornecidos
        $updateFields = [];
        $params = [];
        
        if (array_key_exists('nome', $input)) {
            $updateFields[] = 'nome = ?';
            $params[] = $input['nome'] === '' ? null : $input['nome'];
        }
        
        if (array_key_exists('descricao', $input)) {
            $updateFields[] = 'descricao = ?';
            $params[] = $input['descricao'] === '' ? null : $input['descricao'];
        }
        
        if (array_key_exists('valor', $input)) {
            $updateFields[] = 'valor = ?';
            $params[] = $input['valor'] === '' ? null : $input['valor'];
        }
        
        if (array_key_exists('ativo', $input)) {
            $updateFields[] = 'ativo = ?';
            $params[] = $input['ativo'] === '' ? null : $input['ativo'];
        }
        
        if (array_key_exists('palavras_chave', $input)) {
            $updateFields[] = 'palavras_chave = ?';
            $params[] = $input['palavras_chave'] === '' ? null : $input['palavras_chave'];
        }
        
        if (array_key_exists('categoria', $input)) {
            $updateFields[] = 'categoria = ?';
            $params[] = $input['categoria'] === '' ? null : $input['categoria'];
        }
        
        if (array_key_exists('observacoes', $input)) {
            $updateFields[] = 'observacoes = ?';
            $params[] = $input['observacoes'] === '' ? null : $input['observacoes'];
        }
        
        if (array_key_exists('preparo_necessario', $input)) {
            $updateFields[] = 'preparo_necessario = ?';
            $params[] = $input['preparo_necessario'] === '' ? null : $input['preparo_necessario'];
        }
        
        if (array_key_exists('anestesia_tipo', $input)) {
            $updateFields[] = 'anestesia_tipo = ?';
            $params[] = $input['anestesia_tipo'] === '' ? null : $input['anestesia_tipo'];
        }
        
        if (array_key_exists('local_realizacao', $input)) {
            $updateFields[] = 'local_realizacao = ?';
            $params[] = $input['local_realizacao'] === '' ? null : $input['local_realizacao'];
        }
        
        if (empty($updateFields)) {
            throw new Exception('Nenhum campo para atualizar foi fornecido');
        }
        
        // Adicionar updated_at e ID para WHERE
        $updateFields[] = 'updated_at = NOW()';
        $params[] = $input['id']; // ID para WHERE
        
        $sql = "UPDATE servicos_clinica SET " . implode(', ', $updateFields) . " WHERE id = ?";
        $stmt = $pdo->prepare($sql);
        $stmt->execute($params);
        
        echo json_encode([
            'success' => true,
            'message' => 'Serviço atualizado com sucesso'
        ]);
        
    } elseif ($_SERVER['REQUEST_METHOD'] === 'DELETE') {
        // Deletar serviço
        $input = json_decode(file_get_contents('php://input'), true);
        
        if (!$input || !isset($input['id'])) {
            throw new Exception('ID do serviço é obrigatório');
        }
        
        $stmt = $pdo->prepare("DELETE FROM servicos_clinica WHERE id = ?");
        $stmt->execute([$input['id']]);
        
        if ($stmt->rowCount() === 0) {
            throw new Exception('Serviço não encontrado');
        }
        
        echo json_encode([
            'success' => true,
            'message' => 'Serviço deletado com sucesso'
        ]);
        
    } else {
        throw new Exception('Método não permitido');
    }
    
} catch (Exception $e) {
    echo json_encode([
        'success' => false,
        'message' => $e->getMessage()
    ]);
} 

File: app/routers/panel/servicos.py
from fastapi import APIRouter, Depends, HTTPException, Request
from fastapi.responses import JSONResponse
from fastapi.encoders import jsonable_encoder
from ...core.db import get_db, is_postgres_connection
from ..auth import verify_admin_user
from ...schemas.panel import ServicoCreate, ServicoUpdate

router = APIRouter(prefix="/panel", tags=["panel-servicos"], dependencies=[Depends(verify_admin_user)])

@router.get("/servicos")
async def listar_servicos(request: Request):
    try:
        db_gen = get_db()
        db = next(db_gen)
        try:
            id_str = request.query_params.get("id")
            with db.cursor() as cur:
                if id_str is not None:
                    cur.execute("SELECT * FROM servicos_clinica WHERE id = %s", (id_str,))
                    servico = cur.fetchone()
                    if servico:
                        return JSONResponse(content={"success": True, "servico": servico})
                    return JSONResponse(content={"success": False, "message": "Serviço não encontrado"})
                cur.execute("SELECT * FROM servicos_clinica ORDER BY nome")
                servicos = cur.fetchall()
                return JSONResponse(content=jsonable_encoder({"success": True, "servicos": servicos}))
        finally:
            try:
                db.close()
            except Exception:
                pass
    except Exception as e:
        return JSONResponse(content={"success": False, "message": "Erro ao buscar serviços", "error": str(e)}, status_code=500)


@router.post("/servicos")
async def criar_servico(payload: ServicoCreate):
    try:
        db_gen = get_db()
        db = next(db_gen)
        try:
            with db.cursor() as cur:
                if is_postgres_connection(db):
                    cur.execute(
                        """
                        INSERT INTO servicos_clinica (
                            nome, descricao, valor, categoria, palavras_chave, observacoes, 
                            preparo_necessario, anestesia_tipo, local_realizacao, ativo
                        ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s) RETURNING id
                        """,
                        (
                            payload.nome, payload.descricao, payload.valor, payload.categoria,
                            payload.palavras_chave, payload.observacoes, payload.preparo_necessario,
                            payload.anestesia_tipo, payload.local_realizacao, 1 if payload.ativo else 0
                        )
                    )
                    result = cur.fetchone()
                    servico_id = result["id"] if isinstance(result, dict) else result[0]
                else:
                    cur.execute(
                        """
                        INSERT INTO servicos_clinica (
                            nome, descricao, valor, categoria, palavras_chave, observacoes, 
                            preparo_necessario, anestesia_tipo, local_realizacao, ativo
                        ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
                        """,
                        (
                            payload.nome, payload.descricao, payload.valor, payload.categoria,
                            payload.palavras_chave, payload.observacoes, payload.preparo_necessario,
                            payload.anestesia_tipo, payload.local_realizacao, 1 if payload.ativo else 0
                        )
                    )
                    servico_id = cur.lastrowid
                
                return JSONResponse(content={"success": True, "message": "Serviço criado com sucesso", "id": servico_id})
        finally:
            try:
                db.close()
            except Exception:
                pass
    except Exception as e:
        return JSONResponse(content={"success": False, "message": f"Erro ao criar serviço: {str(e)}"}, status_code=500)


@router.put("/servicos/{servico_id}")
async def atualizar_servico(servico_id: int, payload: ServicoUpdate):
    try:
        db_gen = get_db()
        db = next(db_gen)
        try:
            # Montar query dinamicamente baseada nos campos fornecidos
            updates = []
            values = []
            
            for field, value in payload.dict(exclude_unset=True).items():
                if value is not None:
                    updates.append(f"{field} = %s")
                    if field == "ativo":
                        values.append(1 if value else 0)
                    else:
                        values.append(value)
            
            if not updates:
                return JSONResponse(content={"success": False, "message": "Nenhum campo para atualizar"}, status_code=400)
            
            # Adicionar updated_at
            updates.append("updated_at = CURRENT_TIMESTAMP")
            values.append(servico_id)
            
            with db.cursor() as cur:
                # Verificar se existe
                cur.execute("SELECT id FROM servicos_clinica WHERE id = %s", (servico_id,))
                if not cur.fetchone():
                    return JSONResponse(content={"success": False, "message": "Serviço não encontrado"}, status_code=404)
                
                # Atualizar
                query = f"UPDATE servicos_clinica SET {', '.join(updates)} WHERE id = %s"
                cur.execute(query, values)
                
                return JSONResponse(content={"success": True, "message": "Serviço atualizado com sucesso"})
        finally:
            try:
                db.close()
            except Exception:
                pass
    except Exception as e:
        return JSONResponse(content={"success": False, "message": f"Erro ao atualizar serviço: {str(e)}"}, status_code=500)


@router.delete("/servicos/{servico_id}")
async def deletar_servico(servico_id: int):
    try:
        db_gen = get_db()
        db = next(db_gen)
        try:
            with db.cursor() as cur:
                # Verificar se existe
                cur.execute("SELECT id FROM servicos_clinica WHERE id = %s", (servico_id,))
                if not cur.fetchone():
                    return JSONResponse(content={"success": False, "message": "Serviço não encontrado"}, status_code=404)
                
                # Desativar em vez de deletar (soft delete)
                cur.execute("UPDATE servicos_clinica SET ativo = 0 WHERE id = %s", (servico_id,))
                
                return JSONResponse(content={"success": True, "message": "Serviço desativado com sucesso"})
        finally:
            try:
                db.close()
            except Exception:
                pass
    except Exception as e:
        return JSONResponse(content={"success": False, "message": f"Erro ao deletar serviço: {str(e)}"}, status_code=500)


@router.get("/servicos/{servico_id}")
async def obter_servico(servico_id: int):
    try:
        db_gen = get_db()
        db = next(db_gen)
        try:
            with db.cursor() as cur:
                cur.execute("SELECT * FROM servicos_clinica WHERE id = %s", (servico_id,))
                servico = cur.fetchone()
                if not servico:
                    return JSONResponse(content={"success": False, "message": "Serviço não encontrado"}, status_code=404)
                
                return JSONResponse(content={"success": True, "servico": servico})
        finally:
            try:
                db.close()
            except Exception:
                pass
    except Exception as e:
        return JSONResponse(content={"success": False, "message": f"Erro ao buscar serviço: {str(e)}"}, status_code=500)


# Endpoints para compatibilidade com o frontend atual
@router.put("/servicos")
async def atualizar_servico_legacy(payload: dict):
    if not payload or "id" not in payload:
        raise HTTPException(status_code=400, detail="ID é obrigatório")
    
    servico_id = payload.pop("id")
    # Converter valores vazios para None
    for key, value in payload.items():
        if value == "":
            payload[key] = None
    
    servico_update = ServicoUpdate(**payload)
    return await atualizar_servico(servico_id, servico_update)


@router.delete("/servicos")
async def deletar_servico_legacy(payload: dict):
    if not payload or "id" not in payload:
        raise HTTPException(status_code=400, detail="ID do serviço é obrigatório")
    
    return await deletar_servico(payload["id"])